app/code/Magento/PaypalGraphQl/Observer/PayflowProSetCcData.php

```diff
@@ -0,0 +1,99 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+
+declare(strict_types=1);
+
+namespace Magento\PaypalGraphQl\Observer;
+
+use Magento\Customer\Model\Session as CustomerModelSession;
+use Magento\Framework\App\Config\ScopeConfigInterface;
+use Magento\Framework\Event\Observer;
+use Magento\Framework\GraphQl\Exception\GraphQlInputException;
+use Magento\Payment\Observer\AbstractDataAssignObserver;
+use Magento\Quote\Api\Data\PaymentInterface;
+
+/**
+ * Class PayflowProSetCcData set CcData to quote payment
+ * @SuppressWarnings(PHPMD.CookieAndSessionMisuse)
+ */
+class PayflowProSetCcData extends AbstractDataAssignObserver
+{
+    const XML_PATH_PAYMENT_PAYFLOWPRO_CC_VAULT_ACTIVE = "payment/payflowpro_cc_vault/active";
+    const IS_ACTIVE_PAYMENT_TOKEN_ENABLER = "is_active_payment_token_enabler";
+
+    /**
+     * @var CustomerModelSession
+     */
+    private $customerSession;
+
+    /**
+     * Core store config
+     *
+     * @var ScopeConfigInterface
+     */
+    private $scopeConfig;
+
+    /**
+     * @param CustomerModelSession $customerSession
+     * @param ScopeConfigInterface $scopeConfig
+     */
+    public function __construct(
+        CustomerModelSession $customerSession,
+        ScopeConfigInterface $scopeConfig
+    ) {
+        $this->customerSession = $customerSession;
+        $this->scopeConfig = $scopeConfig;
+    }
+
+    /**
+     * Set CcData
+     *
+     * @param Observer $observer
+     *
+     * @throws GraphQlInputException
+     */
+    public function execute(Observer $observer)
+    {
+        $dataObject = $this->readDataArgument($observer);
+        $additionalData = $dataObject->getData(PaymentInterface::KEY_ADDITIONAL_DATA);
+        $paymentModel = $this->readPaymentModelArgument($observer);
+
+        if (!isset($additionalData['cc_details'])) {
+            return;
+        }
+
+        if ($this->customerSession->isLoggedIn() && $this->isPayflowProVaultEnable()) {
```

Comment:
Hi @Usik2203. Relying on session is discouraged at GraphQl Layer. One possible alternative is to have a validatorpool specific to a payment method, similar to the additionalDataprovider pool and perform validation at the resolver level. You can pass in context and args for the validations.

Reply:
**Done**